src/analysis/ReachingDefinitions/Srg/IntervalMap.h

```diff
@@ -0,0 +1,214 @@
+#ifndef _DG_INTERVALSET_H
+#define _DG_INTERVALSET_H
+
+#include "analysis/Offset.h"
+#include "analysis/ReachingDefinitions/RDMap.h"
+
+namespace dg {
+namespace analysis {
+namespace rd {
+namespace srg {
+namespace detail {
+
+/**
+ * represents a both-side closed interval of values
+ */
+class Interval {
+    using T = Offset;
+
+    T start;
+    T len;
+
+    static inline T min(T a, T b) {
+        return a < b ? a : b;
+    }
+
+    static inline T max(T a, T b) {
+        return a > b ? a : b;
+    }
+
+public:
+    Interval(T start, T len): start(start), len(len) {}
+
+    bool isUnknown() const {
+        return start.isUnknown() || len == 0;
+    }
+
+    bool overlaps(const Interval& other) const {
+        if (isUnknown() || other.isUnknown())
+            return false;
+        return intervalsOverlap(start.offset, len.offset, other.start.offset, other.len.offset);
+    }
+
+    bool isSubsetOf(const Interval& other) const {
+        return start >= other.start && start + len <= other.start + other.len;
+    }
+
+    bool unite(const Interval& other) {
+        if (isUnknown() || other.isUnknown())
+            return false;
+        if (overlaps(other) || start + len == other.start || other.start + other.len == start) {
+            T a = min(start, other.start);
+            T b = max(start + len, other.start + other.len);
+            len = b - a;
+            start = a;
+            return true;
+        }
+        return false;
+    }
+
+    T getStart() const {
+        return start;
+    }
+    
+    T getLength() const {
+        return len;
+    }
+
+};
+
+class DisjointIntervalSet {
+    std::vector<Interval> intervals;
+public:
+
+    DisjointIntervalSet() { }
+
+    template <typename C>
+    DisjointIntervalSet(const C& c) {
+        for (const Interval& interval : c) {
+            insert(interval);
+        }
+    }
+
+    void insert(Interval interval) {
+        // find & remove all overlapping intervals
+        for (auto it = begin(); it != end(); ) {
+            if (interval.unite(*it)) {
+                it = intervals.erase(it);
+            } else {
+                ++it;
+            }
+        }
+
+        intervals.push_back(std::move(interval));
+    }
+
+    auto cbegin() const -> decltype(intervals.begin()) {
```

Comment:
Isn't `auto` return type  C++14 feature? We use C++11 standard (as defined in CMakeList.txt), so I'm not sure if it won't break anything. Or maybe it is time to switch to a newer standard?

Reply:
The compiler shows no errors, so there should be no problem. [cppreference page](http://en.cppreference.com/w/cpp/language/auto) says it's part of C++11: `auto function -> return type | (2) | (since C++11)`.

Reply:
Great! I should have checked that myself...